lib/internal/streams/operators.js

```diff
@@ -215,6 +219,53 @@ async function * filter(fn, options) {
   yield* this.map(filterFn, options);
 }
 
+// Specific to provide better error to reduce since the argument is only
+// missing if the stream has no items in it - but the code is still appropriate
+class ReduceAwareErrMissingArgs extends ERR_MISSING_ARGS {
+  constructor() {
+    super('reduce');
+    this.message = 'Reduce of an empty stream requires an initial value';
+  }
+}
+
+async function reduce(reducer, initialValue, options) {
+  if (typeof reducer !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'reducer', ['Function', 'AsyncFunction'], reducer);
+  }
+  let hasInitialValue = arguments.length > 1;
+  // TODO(benjamingr) discuss
```

Comment:
This is a question - if reduce (and others I think) get an already aborted signal they will not start iterating the stream and thus not close it.

I think this is probably the wrong behavior and  the stream should be closed regardless but I want to check what people think?

```js
const ac = new AbortController();
ac.abort(); // or just AbortSignal.abort();

getStringStream().reduce((p, c) => {
  return p.concat(c);
}, 0, { signal: ac.signal })
```

In the current implementation this would not attempt to consume the stream. Which may be unexpected IMO but we are aborting the reduce (and not the stream). I can also check what other languages do if people don't have a clear answer on what they think we should do.

cc @ronag @mcollina

Reply:
I agree, because calling `getStringStream().xyz()` is a clear intent to consume that stream. If that's not desirable, the user simply shouldn't call it. Which is a good practice anyway, as it avoids allocating unnecessary resources.

Aborting shouldn't have side effects. At this point in the code, we already have a side effect (creating the stream) which therefore must be reverted by closing the stream. Then the resulting state is the same as when no abort happened.

Reply:
Ok, changed this behavior to destroy the stream, I'll push fixes to the other methods too where applicable.